design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,324 @@
+for more information.
+
+#### Implementation
+The development work on the feature is expected to include coding,
+testing, integration and documentation.
+
+#### Releases
+- **Alpha**:
+  An initial release of the feature for experimental purposes.
+  Recommended for non-production usages, evaluation or testing.
+
+  The API is considered unstable and may change significantly.
+
+  The period in which a feature can remain in Alpha is limited,
+  assuring features are not piling up without control.
+  See [release stage transition table](#release-stage-transition-table)
+  for more information.
+
+  The feature presence is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **Beta**:
+  The first release that can be evaluated with care in production.
+  Acting as a pre-release, its main objective is to collect feedback
+  from users to assure its usefulness and readiness for graduation.
+  If there is no confidence of usage or usefulness, it may remain in
+  this stage for some time.
+
+  However, the period in which a feature can remain in Beta is limited,
+  assuring features are not piling up without control.
+  See [release stage transition table](#release-stage-transition-table)
+  for more information.
+
+  The API is considered stable with care not to break backward compatibility
+  with previous beta releases.
+  This implies that fields may only be added during this stage,
+  not removed or renamed.
+
+  The feature presence is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **GA**:
+  The feature graduated to general-availability (GA) and is now part of
+  the core features.
+
+  The API is considered stable with care not to break backward compatibility
+  with the previous releases.
+
+  The feature functionality is no longer controlled by a FG.
+
+> **Warning**: A Feature Gate flag is solely intended to control
+> feature lifecycle. It should not be confused and used as a cluster
+> configurable enablement of the functionality.
+> In cases where the cluster admin should control a functionality,
+> regardless to the feature stage, dedicated configuration field/s
+> should be included.
+
+#### Removal
+When and if a feature is targeted for deprecation and retirement,
+it needs to pass a deprecation process, depending on its current
+release stage (Alpha, Beta, GA).
+
+For more details, see [here](#deprecation-and-removal).
+
+#### Release Stage Transition Table
+The following table summarized the different release stages with their
+transition requirements and restrictions.
+
+| Stage          | Period range    | F.Gate | Removal Availability       |
+|----------------|-----------------|--------|----------------------------|
+| Alpha          | 1 to 2 releases | YES    | Between **minor** releases |
+| Beta           | 1 to 3 releases | YES    | Between **minor** releases |
```

Comment:
This is no longer correct, FG are disabled by default on Beta as well.

Reply:
Github review mechanism should apologize, not you :)
Very easy to miss comments on threads that are no longer visible or marked resolved.

I think [this](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages) section changed again.

Now it is clearer and states `Usually enabled by default. Beta API groups are disabled by default`.

So it leaves the beta default up to the feature owner. Strange.

I do not really understand why it makes sense to have it by default enabled, to me it will break users if it gets dropped.
Based on this clarification, this can be indeed changed.

I just want to note that when we used a Beta feature from K8S (i.e. depend on it) we usually added a FG and considered it by default disabled. Just because we could not commit about it getting graduated.
That logic tells me that we should keep them by default disabled in our case as well.

Reply:
Another point that is worth mentioning: We do not really have a way to define a FG by default enabled and allow users to override it with explicit disable.
If this process will define such a thing, an implementation will gave to create a new API it.

Reply:
Yup great point, I'd like us to mirror k8s whenever possible but I can't think of way of doing it in this case without rewriting our implementation of featuregates.

I'll drop this for now, disabled by default for new alpha and beta feature gates is fine.

Reply:
> I do not really understand why it makes sense to have it by default enabled, to me it will break users if it gets dropped.

I think that the rationale is that on the one hand the feature is considered stable, but on the other hand they grant you an "escape hatch" if something bad happens. But yeah, I get your point.

> Another point that is worth mentioning: We do not really have a way to define a FG by default enabled and allow users to override it with explicit disable.

That's interesting.
Maybe we'd want to change that sometime.

> I'll drop this for now, disabled by default for new alpha and beta feature gates is fine.

Agreed.

Reply:
What do we need to resolve this conversation?